feat: Create ViewModel for the TrackSessionsActivity

dr0pdb opened the pull request:
Fixes #2053  

Changes: Added ViewModel for the TrackSessionsActivity.
I have implemented the search query inside the ViewModel now, but since the RecyclerView adapter is shared with ```LocationActivity```, i can't delete the filter method inside the Adapter right now. It will be delete in the PR for ```LocationActivity```. 

Screenshots for the change: NA 

@iamareebjamal @the-dagger Please review.

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/viewmodels/TrackSessionsActivityViewModel.java:
```diff
@@ -0,0 +1,131 @@
+            setSearchText(searchText);
+        }
+        return trackSessionLiveData;
+    }
```
@srv-twry Please update the PR

dr0pdb commented:
@iamareebjamal There is one issue. The Activity needs mainly two objects , the Tracks object and a ```List<Session>```. The ```List<Session>``` should also implement filtering. The ```List<Session>``` is fetched from the tracks object using ```track.getSessions()```. So everytime the tracks object changes inside the database we need to update the ```List<Session>``` also.
Considering all these constraints, here is how i have implemented these.

``` Java
    public LiveData<Track> getTracks(int trackId) {
        if (trackLiveData == null || this.trackId != trackId) {
            this.trackId = trackId;
            track = realmRepo.getTrack(trackId);
            filteredSessionsLiveData = new FilterableRealmLiveData<>(track.getSessions().sort("startsAt"));
            track.addChangeListener((RealmObjectChangeListener<Track>) (track, objectChangeSet) -> {
                filterData(searchText);
            });
            trackLiveData = RealmDataRepository.asLiveDataForObject(track);
        }
        return trackLiveData;
    }

    private void filterData(String searchText) {
        setSearchText(searchText);
        final String query = searchText.toLowerCase(Locale.getDefault());

        Predicate<Session> predicate = session -> session.getTitle()
                .toLowerCase(Locale.getDefault())
                .contains(query);

        filteredSessionsLiveData.filter(predicate);
        calculateUpcomingOngoingPosition();
    }

    public LiveData<List<Session>> getSessions(String searchText) {
        filterData(searchText);
        return filteredSessionsLiveData;
    }
```
Now the problem is finding the upcoming and ongoing positions which needs the contents of the ```List<Session>```. There are two ways:
1. Either i move the logic for finding the upcoming and ongoing positions to the View i.e. Activity.
2. Use MutableLiveData and get it's value inside the ViewModel and then find it inside the ViewModel. In that case i won't be able to use FilterableRealmLiveData and hence i would have to implement filtering on the MutableLiveData or create a wrapper for it also.
What do you think is the proper way to handle this ?

iamareebjamal commented:
The first time `getTracks` is called, save the reference of LiveData in a separate variable, this is will give you the reference to track which has all the sessions. Now you can find out the upcoming session through this

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/activities/TrackSessionsActivity.java:
```diff
@@ -120,54 +118,46 @@ protected void onCreate(Bundle savedInstanceState) {
-                sessionsRecyclerView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
+    private void loadTracks() {
+        trackSessionsActivityViewModel.getTrack(trackId).observe(this, track -> {
+            if (track != null) {
```
Don't send track from ViewModel if it was null

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/viewmodels/TrackSessionsActivityViewModel.java:
```diff
@@ -0,0 +1,126 @@
+
+    private void filterData(String searchText) {
+        if (!this.searchText.equals(searchText) || firstTime) {
+            firstTime = false;
```
Why we need firstTime

dr0pdb replied:
Yes, we don't need it. I was getting a blank screen so i tried to avoid it. I found the actual bug now, it was actually in my implementation of the FilterableRealmLiveData.

dr0pdb committed:
feat: Add ViewModel in TrackSessionsActivity
Fixed FilterableRealmLiveData

dr0pdb commented:
Changes: Removed null check and firstTime boolean variable.
Fixed FilterableRealmLiveData to set it's value to the unfiltered data in the constructor.

@iamareebjamal Please review.
